pkg/middleware/stored_session.go

```diff
@@ -105,51 +106,45 @@ func (s *storedSessionLoader) getValidatedSession(rw http.ResponseWriter, req *h
 	return session, nil
 }
 
-// refreshSessionIfNeeded will attempt to refresh a session if the session
-// is older than the refresh period.
+// refreshSessionIfNeeded will attempt to refresh a session
 // It is assumed that if the provider refreshes the session, the session is now
 // valid.
 // If the session requires refreshing but the provider does not refresh it,
 // we must validate the session to ensure that the returned session is still
 // valid.
 func (s *storedSessionLoader) refreshSessionIfNeeded(rw http.ResponseWriter, req *http.Request, session *sessionsapi.SessionState) error {
-	if s.refreshPeriod <= time.Duration(0) || session.Age() < s.refreshPeriod {
-		// Refresh is disabled or the session is not old enough, do nothing
+	if session == nil {
+		// Refresh is disabled or there is no session, do nothing
 		return nil
 	}
 
-	logger.Printf("Refreshing %s old session cookie for %s (refresh after %s)", session.Age(), session, s.refreshPeriod)
-	refreshed, err := s.refreshSessionWithProvider(rw, req, session)
-	if err != nil {
-		return err
-	}
-
-	if !refreshed {
-		// Session wasn't refreshed, so make sure it's still valid
+	if session.RefreshToken == "" || !s.refresh.ShouldRefresh(time.Now(), *session.CreatedAt, *session.ExpiresOn) {
+		// Session cannot be refreshed due to absent refresh token or has not yet expired
+		// Validate existing session
 		return s.validateSession(req.Context(), session)
```

Comment:
Hmmm - we need to think through this logic and be very careful. Some providers in their `ValidateSession` logic query an external `ValidateURL`. At the moment, I think this logic results in that happening for every request before the session needs refreshing, which is not good.

@JoelSpeed - Long term, what use do we see for `ValidateSession`? It historically was only called if timing options asked for a session refresh, but the provider didn't choose to refresh. So we fell back to `ValidateSession` to at least make sure the Access/ID Tokens were still fine in the session.

If we always opportunistically refresh sessions early if a user configures it that way, do we ever need to `ValidateSession`...? 🤔

Reply:
So this might be a codepath we lost for a lot of providers if we changed `RefreshSessionIfNeeded` to not return `(bool, error)` anymore and instead return `error`:

The default implementation for `RefreshSessionIfNeeded` returned `false, nil`. Which would lead to all providers that didn't implement `RefreshSessionIfNeeded` at this point to call `ValidateSession`.

If a provider didn't implement `ValidateSession` there is a default implementation of that which does use the access token to hit a provider validateURL. We are losing that codepath in this implementation which needs to be fixed I think.

Reply:
Now, for those providers in that boat:

Do they keep hitting a validateURL for every request in the time gap between a configured refresh time & expiration? That is not good if so (but I think might be the case today... 😨 ).

We need to think about potentially how we manage last validation time, it might need to be a time field we track on sessions.

Reply:
Let me take a stab to try to restore this code path to be as close as possible to what was there.